crates/ra_assists/src/handlers/replace_qualified_name_with_use.rs

```diff
@@ -73,6 +89,59 @@ fn collect_hir_path_segments(path: &hir::Path) -> Option<Vec<SmolStr>> {
     Some(ps)
 }
 
+/// Adds replacements to `re` that shorten `path` in all descendants of `node`.
+fn shorten_paths(re: &mut SyntaxRewriter<'static>, node: SyntaxNode, path: &ModPath) {
+    for child in node.children() {
+        match_ast! {
+            match child {
+                // Don't modify `use` items, as this can break the `use` item when injecting a new
+                // import into the use tree.
+                ast::UseItem(_it) => continue,
+                // Don't descend into submodules, they don't have the same `use` items in scope.
+                ast::Module(_it) => continue,
+
+                ast::Path(p) => {
+                    match maybe_replace_path(re, &p, path) {
+                        Some(()) => {},
+                        None => shorten_paths(re, p.syntax().clone(), path),
+                    }
+                },
+                _ => shorten_paths(re, child, path),
+            }
+        }
+    }
+}
+
+fn maybe_replace_path(
+    re: &mut SyntaxRewriter<'static>,
+    p: &ast::Path,
+    path: &ModPath,
+) -> Option<()> {
+    let hir_path = hir::Path::from_ast(p.clone())?;
```

Comment:
`hir::Path::from_ast` should really die in flames. It's a broken API, because it can't take hygiene into account. There's only a couple of calls to this function from IDE, and we probably should find ways to remove them.